### ISSUE TITLE
New configuration system spec draft

### ISSUE THREAD

**Issue by danielballan** [bluesky/databroker#194]
Goal: Support multiple databrokers in the same process so it is easy to switch between beamlines or combine data between beamlines or juggle different databrokers for raw and analyzed data.

Plan, previously outlined with @tacaswell and @stuartcampbell on a whiteboard:

Deprecate `databroker.DataBroker` and `databroker.db` with noisy warnings. (We probably have to maintain them for a long time or we will break ~1000 user notebooks.)

The new encouraged usage will be:

```python
from databroker import Broker
db = Broker.named('chx')
```

relying on a config file something like:

```yaml
# /etc/databroker/chx.yml
headersource:  # mapped to one dict
  module: databroker.headersource.mongo
  class: MDSRO
  config:
    database: metdatastore-production-v1
    host: xf23id-broker
    port: 27017
    timezone: US/Eastern
    
event_sources:  # mapped to a list of dicts (in this case, a list with one element)
  - module: databroker.headersource.mongo   # in the future, this will a proper EventSource
    class: MDSRO
    config:
      database: metdatastore-production-v1
      host: xf23id-broker
      port: 27017
      timezone: US/Eastern
    filestore:
      module: databroker.filestore.mongo
      class: FileStoreRO
      config:
        database: filestore-production-v1
        host: xf23id-broker
        port: 27017
```

New broker code would look something like:

```python
SEARCH_PATH = ('/etc/databroker', os.path.expanduser('~/.config/databroker/'))

def lookup_config(name):
    # Look for a file called {name}.yml in the SEARCH_PATH.
    # Open and parse it into a dict
    return config

from importlib import import_module

class Broker:
    ...
    @classmethod
    def from_config(cls, config):
        modname = config['headersource']['module']  # a string
        clsname = config['headersource']['class']  # a string
        config = config['headersource']['config']  # a dict
        mod = import_module(modname)
        cls = getattr(mod, clsname)
        headersource = cls(config)
        # similar for event_source(s) and their filestores
        return cls(headersource, *eventsources)

    @classmethod
    def named(cls, name):
        return cls.from_config(lookup_config(name))
```

**Comment by CJ-Wright** [comment]
Does this mean that if we stuff the `name` of the databroker into the documents we could reference/get data from that broker without having to have that broker loaded ahead of time? This would fix some of the problems on my side. Eg I'd like to use raw data stored in filestore in my analyzed database (I ran a huge filtering algorithm on it and want to persist the output) but don't want to write down a separate copy per databroker.

**Comment by danielballan** [comment]
Closed by #218
